File: trunk/CharasLike/Classes/Character.vb
Public Enum EFlipFlag
    FlipNone = 0
    FlipHorizontal = 1
    FlipVertical = 2
End Enum

Public Class RPGLayer
    Private Offset_ As New Point
    Public Property Offset() As Point
        Get
            Return Offset_
        End Get
        Set(ByVal value As Point)
            Offset_ = value
        End Set
    End Property

    Private Image_ As RPGImage
    Public Property Image() As RPGImage
        Get
            Return Image_
        End Get
        Set(ByVal value As RPGImage)
            Image_ = value
        End Set
    End Property

    Private Name_ As String
    Public Property Name() As String
        Get
            Return Name_
        End Get
        Set(ByVal value As String)
            Name_ = value
        End Set
    End Property

    Private Matrix_ As New QColorMatrix
    Public Property Matrix() As QColorMatrix
        Get
            Return Matrix_
        End Get
        Set(ByVal value As QColorMatrix)
            Matrix_ = value
        End Set
    End Property

    Private MatrixValues_(3) As Single
    Public Sub SetMatrixValue(ByVal index As Integer, ByVal value As Single)
        MatrixValues_(index) = value
    End Sub
    Public Function GetMatrixValue(ByVal index As Integer)
        Return MatrixValues_(index)
    End Function

    Private Alpha_ As Integer
    Public Property Alpha() As Integer
        Get
            Return Alpha_
        End Get
        Set(ByVal value As Integer)
            Alpha_ = value
        End Set
    End Property

    Private SubLayers_ As New Collections.ObjectModel.Collection(Of RPGLayer)
    Public ReadOnly Property SubLayers() As Collections.ObjectModel.Collection(Of RPGLayer)
        Get
            Return SubLayers_
        End Get
    End Property

    Private Parent_ As RPGLayer
    Public Property Parent() As RPGLayer
        Get
            Return Parent_
        End Get
        Set(ByVal value As RPGLayer)
            Parent_ = value
        End Set
    End Property

    Private OwnedCharacter_ As RPGCharacter
    Public Property OwnedCharacter() As RPGCharacter
        Get
            Return OwnedCharacter_
        End Get
        Set(ByVal value As RPGCharacter)
            OwnedCharacter_ = value
        End Set
    End Property

    Private FlipFlags_ As Integer
    Public Property FlipFlags() As Integer
        Get
            Return FlipFlags_
        End Get
        Set(ByVal value As Integer)
            FlipFlags_ = value
        End Set
    End Property

    Private FlippedImage_ As Bitmap
    Public Property FlippedImage() As Bitmap
        Get
            Return FlippedImage_
        End Get
        Set(ByVal value As Bitmap)
            FlippedImage_ = value
        End Set
    End Property

    Public Sub New()
        Offset = Point.Empty
        Image = Nothing
        Name = ""
        Parent = Nothing
        Matrix.Reset()
        Alpha = 255
        FlipFlags = EFlipFlag.FlipNone
        FlippedImage = Nothing

        SetMatrixValue(0, 0)
        SetMatrixValue(1, 1)
        SetMatrixValue(2, 1)
        SetMatrixValue(3, 0)
    End Sub

    Public Sub SetImage(ByVal Img As RPGPictureBox)
        If Img Is Nothing Then
            Image = Nothing
            LayersWindow.PictureBox1.Image = Nothing
            FlippedImage = Nothing
            Return
        End If

        Image = Img.MainImage
        LayersWindow.PictureBox1.Image = Img.MyImage
        UpdateFlippedImage()
    End Sub

    Public Sub UpdateFlippedImage()
        FlippedImage = Image.RawBitmap.Clone()
        If (FlipFlags And (EFlipFlag.FlipHorizontal Or EFlipFlag.FlipVertical)) = (EFlipFlag.FlipHorizontal Or EFlipFlag.FlipVertical) Then
            FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipXY)
        ElseIf (FlipFlags And EFlipFlag.FlipHorizontal) = EFlipFlag.FlipHorizontal Then
            FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipX)
        ElseIf (FlipFlags And EFlipFlag.FlipVertical) = EFlipFlag.FlipVertical Then
            FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipY)
        End If
    End Sub

    Public Sub Save(ByVal File As IO.BinaryWriter, ByVal ImagesToo As Boolean)
        File.Write(Offset.X)
        File.Write(Offset.Y)
        File.Write(Name)
        File.Write(Image IsNot Nothing)
        If Image IsNot Nothing Then
            File.Write(Image.Path)

            If ImagesToo Then
                File.Write(Image.OriginalBitmap.Width)
                File.Write(Image.OriginalBitmap.Height)

                Canvas.WriteBitmapToFile(File, Image.OriginalBitmap)
            End If

        End If

        For x As Integer = 0 To 4
            For y As Integer = 0 To 4
                File.Write(Matrix.MyMatrix(x, y))
            Next
        Next
        For i As Integer = 0 To 3
            File.Write(MatrixValues_(i))
        Next
        File.Write(Alpha)
        File.Write(FlipFlags)

        ' Parent is not saved/loaded, it is re-created on run-time
        File.Write(SubLayers.Count)
        For Each SubLayer In SubLayers
            SubLayer.Save(File, ImagesToo)
        Next
    End Sub

    Public Sub Load(ByVal File As IO.BinaryReader, ByVal ImagesToo As Boolean)
        Offset = New Point(File.ReadInt32(), File.ReadInt32())
        Name = File.ReadString()
        If File.ReadBoolean() = True Then
            Dim Img As String = File.ReadString()
            Image = Images.FindImage(Img)
            If Image IsNot Nothing Then FlippedImage = Image.RawBitmap

            If ImagesToo Then
                Dim Width As Integer = File.ReadInt32(), Height As Integer = File.ReadInt32()

                Using BMP As New Bitmap(Width, Height, Imaging.PixelFormat.Format32bppArgb)
                    Canvas.ReadBitmapFromFile(File, Width, Height, BMP)
                    ' Do we have this image?
                    If IO.File.Exists(Img) Then
                        If Images.BitmapCompare(BMP, Images.FindImage(Img).OriginalBitmap) = False Then
                            ' Message the user
                            If MsgBox("The image " + Img + " already exists in your image folder, but yours differs from the one in the file. Would you like to overwrite with this one?", MsgBoxStyle.YesNo) = MsgBoxResult.Yes Then
                                BMP.Save(Img)
                                Canvas.ImagesAcquired.Add(Img)
                            End If
                        End If
                    ElseIf IO.File.Exists(Img) = False Then
                        IO.Directory.CreateDirectory(Img.Substring(0, Img.LastIndexOf("\")))

                        ' Save the bitmap
                        BMP.Save(Img)
                        Canvas.ImagesAcquired.Add(Img)

                        ' Set our image to a temporary image for later re-loading
                        Image = New RPGImage
                        Image.Path = Img
                    End If
                End Using
            Else
                If Image Is Nothing Then
                    Canvas.PackImagesNotFound.Add(Img)
                End If
            End If
        End If

        For x As Integer = 0 To 4
            For y As Integer = 0 To 4
                Matrix.MyMatrix(x, y) = File.ReadSingle()
            Next
        Next
        For i As Integer = 0 To 3
            MatrixValues_(i) = File.ReadSingle()
        Next
        Alpha = File.ReadInt32()
        FlipFlags = File.ReadInt32()

        Dim SubLayerNum As Integer = File.ReadInt32()
        For i As Integer = 0 To SubLayerNum - 1
            Dim SubLayer As New RPGLayer
            SubLayer.Load(File, ImagesToo)
            SubLayer.Parent = Me
            SubLayers.Add(SubLayer)
        Next
    End Sub

    Public Function GenerateColorMatrix()
        Dim Attr As New Imaging.ImageAttributes
        Attr.SetColorMatrix(Matrix.MyMatrix, Imaging.ColorMatrixFlag.Default, Imaging.ColorAdjustType.Bitmap)
        Return Attr
    End Function

    Public Sub ReloadImagesSubLayers(ByVal Restore As Boolean)
        If Restore = False Then
            For Each SubLayer In SubLayers
                If SubLayer.Image IsNot Nothing Then
                    Dim Img As RPGImage = SubLayer.Image

                    SubLayer.Image = New RPGImage
                    SubLayer.Image.Path = Img.Path
                End If
                SubLayer.ReloadImagesSubLayers(Restore)
            Next
        Else
            For Each SubLayer In SubLayers
                If SubLayer.Image IsNot Nothing Then
                    SubLayer.Image = Images.FindImage(SubLayer.Image.Path)
                    SubLayer.UpdateFlippedImage()
                End If
                SubLayer.ReloadImagesSubLayers(Restore)
            Next
        End If
    End Sub

    Public Sub SetColorMatrix()
        Matrix.Reset()
        Matrix.ScaleColors(Single.Parse(LayersWindow.TextBox4.Text))
        Matrix.TranslateColors(Single.Parse(LayersWindow.TextBox6.Text))
        Matrix.SetSaturation(Single.Parse(LayersWindow.TextBox3.Text))
        Matrix.RotateHue(Integer.Parse(LayersWindow.TextBox2.Text))
        Matrix.ScaleOpacity(Alpha / 255)
        Canvas.UpdateDrawing()
    End Sub

    ' CANVAS RENDERING
    Public Sub RenderCanvasLayer(ByVal gfx As Graphics)
        Dim Zoom As Single = CType(Canvas.ZoomNumber, Single) / 100
        Dim ContentWidth As Integer = CType(Canvas.ToolStripContainer1.ContentPanel.Width / 2, Integer)
        Dim ContentHeight As Integer = CType(Canvas.ToolStripContainer1.ContentPanel.Height / 2, Integer)
        Dim bmp As Bitmap = FlippedImage
        Dim destRect As Rectangle
        Dim srcRect As Rectangle
        Dim Attr As Imaging.ImageAttributes = GenerateColorMatrix()

        ' Frame mode
        If Canvas.FrameAnimationToolStripMenuItem.Checked Then
            Dim BmpWidth As Integer = (((bmp.Width \ Image.GameFile.SheetRows) * Zoom) / 2)
            Dim BmpHeight As Integer = (((bmp.Height \ Image.GameFile.SheetColumns) * Zoom) / 2)
            Dim DrawX As Integer = (ContentWidth - BmpWidth) + (Offset.X * Zoom)
            Dim DrawY As Integer = (ContentHeight - BmpHeight) + (Offset.Y * Zoom)
            Dim Frame As Integer

            If Canvas.AnimationIndex = 0 Then
                Frame = 0
            Else
                Frame = Images.CurrentGameFile.Animations(Canvas.AnimationIndex - 1).GetFrame(Canvas.AnimationFrameIndex)
            End If

            destRect = New Rectangle(DrawX, DrawY, (bmp.Width \ Image.GameFile.SheetRows) * Zoom, (bmp.Height \ Image.GameFile.SheetColumns) * Zoom)
            srcRect = Character.GetRectangleForBitmapFrame(bmp, Frame, Image.GameFile.SheetRows, Image.GameFile.SheetColumns)
            gfx.DrawImage(bmp, destRect, srcRect.X, srcRect.Y, srcRect.Width, srcRect.Height, GraphicsUnit.Pixel, Attr)
        Else ' Sheet mode
            Dim BmpWidth As Integer = ((bmp.Width * Zoom) / 2)
            Dim BmpHeight As Integer = ((bmp.Height * Zoom) / 2)
            Dim DrawX As Integer = (ContentWidth - BmpWidth) + (Offset.X * Zoom) + Canvas.CanvasCamera.X
            Dim DrawY As Integer = (ContentHeight - BmpHeight) + (Offset.Y * Zoom) + Canvas.CanvasCamera.Y

            destRect = New Rectangle(New Point(DrawX, DrawY), New Size(bmp.Width * Zoom, bmp.Height * Zoom))
            srcRect = New Rectangle(0, 0, bmp.Width, bmp.Height)

            gfx.DrawImage(bmp, destRect, srcRect.X, srcRect.Y, srcRect.Width, srcRect.Height, GraphicsUnit.Pixel, Attr)
        End If
    End Sub

    Public Sub DrawCanvasLayers(ByVal e As Graphics)
        If Image IsNot Nothing Then RenderCanvasLayer(e)

        For Each SubLayer In SubLayers
            SubLayer.DrawCanvasLayers(e)
        Next
    End Sub

    Public Sub RenderLayerToGraphics(ByVal gfx As Graphics)
        Dim WImg = OwnedCharacter.SizeOfOutput
        Dim ContentSize As New Size(WImg.Width / 2, WImg.Height / 2)
        Dim BmpSize As New Size(FlippedImage.Width / 2, FlippedImage.Height / 2)
        Dim DrawLocation As Point = (New Point(ContentSize - BmpSize)) + Offset

        gfx.DrawImage(FlippedImage, New Rectangle(DrawLocation, FlippedImage.Size), 0, 0, FlippedImage.Width, FlippedImage.Height, GraphicsUnit.Pixel, GenerateColorMatrix())
    End Sub

    Public Sub RenderLayerToGraphicsRecursion(ByVal e As Graphics)
        If Image IsNot Nothing Then RenderLayerToGraphics(e)
        For Each SubLayer In SubLayers
            SubLayer.RenderLayerToGraphicsRecursion(e)
        Next
    End Sub
End Class

Public Class RPGCharacter
    Private Name_ As String ' Unique ID for character file (for multi-sheet)
    Public Property Name() As String
        Get
            Return Name_
        End Get
        Set(ByVal value As String)
            Name_ = value
        End Set
    End Property

    Private LastSavePath_ As String ' Last place we saved this specific character file
    Public Property LastSavePath() As String
        Get
            Return LastSavePath_
        End Get
        Set(ByVal value As String)
            LastSavePath_ = value
        End Set
    End Property

    Private Layers_ As New Collections.ObjectModel.Collection(Of RPGLayer)
    Public ReadOnly Property Layers() As Collections.ObjectModel.Collection(Of RPGLayer)
        Get
            Return Layers_
        End Get
    End Property

    Public Sub Save(ByVal Writer As IO.BinaryWriter, ByVal ImagesToo As Boolean)
        Writer.Write(Layers.Count)
        For Each Layer In Layers
            Layer.Save(Writer, ImagesToo)
        Next
    End Sub

    Public Sub Load(ByVal Reader As IO.BinaryReader, ByVal ImagesToo As Boolean)
        Dim Count As Integer = Reader.ReadInt32()

        For i As Integer = 0 To Count - 1
            Dim NewLayer As New RPGLayer
            NewLayer.Load(Reader, ImagesToo)
            Layers.Add(NewLayer)
        Next

        For Each Layer In Layers
            Dim Node As New RPGNode
            Node.Layer = Layer
            Node.Text = Layer.Name

            LayersWindow.TreeView1.Nodes.Add(Node)

            Node.RecurseAndLoadSubNodes()
        Next
    End Sub

    Private Function SizeRecursion(ByVal LayerCollection As Collections.ObjectModel.Collection(Of RPGLayer), ByVal CurrentSize As Size)
        For Each Layer In LayerCollection
            CurrentSize = SizeRecursion(Layer.SubLayers, CurrentSize)
            If Layer.FlippedImage Is Nothing Then Continue For
            If Layer.FlippedImage.Size.Width > CurrentSize.Width And Layer.FlippedImage.Size.Height > CurrentSize.Height Then CurrentSize = Layer.FlippedImage.Size
        Next

        Return CurrentSize
    End Function

    Public ReadOnly Property SizeOfOutput() As Size
        Get
            Dim Width As Integer, Height As Integer
            If Layers.Count = 0 Then
                Width = 1
                Height = 1
            Else
                Dim LargestSize As New Size(1, 1)
                Return SizeRecursion(Layers, LargestSize)
            End If
            Return New Size(Width, Height)
        End Get
    End Property

    Public Function CreateLayer(ByVal Node As TreeNode)
        Dim Layer As New RPGLayer

        If Node IsNot Nothing And Node.Parent IsNot Nothing And (Node.Parent IsNot LayersWindow.SplitContainer1.Panel1) Then
            ' Use our parent layer
            Layer.Parent = CType(Node.Parent, RPGNode).Layer
            Layer.OwnedCharacter = Layer.Parent.OwnedCharacter
            Layer.Parent.SubLayers.Add(Layer)
        Else
            Layer.Parent = Nothing
            Layer.OwnedCharacter = Me
            Layers.Add(Layer)
        End If

        Layer.Name = Node.Text
        Return Layer
    End Function

    Public Sub RemoveLayer(ByVal Layer As RPGLayer)
        ' Do we have a parent?
        If Layer.Parent IsNot Nothing Then
            ' Remove us from sublayers
            Layer.Parent.SubLayers.Remove(Layer)
        Else
            Layers.Remove(Layer)
        End If
    End Sub

    Public Sub MoveLayer(ByVal Layer As RPGLayer, ByVal Down As Boolean)
        If Layer.Parent IsNot Nothing Then
            Dim OldIndex As Integer = Layer.Parent.SubLayers.IndexOf(Layer)
            Layer.Parent.SubLayers.Remove(Layer)
            If (Down) Then
                Layer.Parent.SubLayers.Insert(OldIndex + 1, Layer)
            Else
                Layer.Parent.SubLayers.Insert(OldIndex - 1, Layer)
            End If
        Else
            Dim OldIndex As Integer = Character.RPGCharacterFiles(0).Layers.IndexOf(Layer)
            Layers.Remove(Layer)
            If (Down) Then
                Layers.Insert(OldIndex + 1, Layer)
            Else
                Layers.Insert(OldIndex - 1, Layer)
            End If
        End If
        Canvas.UpdateDrawing()
    End Sub
End Class

Public Class Character
    Public Shared RPGCharacterFiles As New Collections.ObjectModel.Collection(Of RPGCharacter)

    Public Shared Function GetRectangleForBitmapFrame(ByVal bmp As Image, ByVal FrameNumber As Integer, ByVal NumRows As Integer, ByVal NumColumns As Integer)
        If bmp Is Nothing Then Return New Rectangle(0, 0, 0, 0)
        Dim SpriteWidth As Integer = bmp.Width \ NumRows, SpriteHeight As Integer = bmp.Height \ NumColumns
        Dim FramesPerRow As Integer = bmp.Width \ SpriteWidth

        Return New Rectangle(((FrameNumber Mod FramesPerRow) * SpriteWidth), ((FrameNumber \ FramesPerRow) * SpriteHeight), SpriteWidth, SpriteHeight)
    End Function
End Class

File: trunk/Classes/Character.cs
using System.Collections.Generic;
using System;
using System.Drawing;
using System.Diagnostics;
using System.Collections;
using System.Windows.Forms;

namespace OpenCharas
{
	public enum EFlipFlag
	{
		FlipNone		= 0,
		FlipHorizontal	= 1,
		FlipVertical	= 2
	}

	public class RPGLayer
	{
		public Point Offset { get; set; }
		public RPGImage Image { get; set; }
		public string Name { get; set; }
		public QColorMatrix Matrix { get; set; }
		public float[] MatrixValues { get; private set; }
		public int Alpha { get; set; }
		public List<RPGLayer> SubLayers { get; private set; }
		public RPGLayer Parent { get; set; }
		public RPGCharacter OwnedCharacter { get; set; }
		public EFlipFlag FlipFlags { get; set; }
		public bool Inverted { get; set; }
		public Bitmap FlippedImage { get; set; }
		public RPGNode Node { get; set; }

		public RPGLayer()
		{
			Offset = Point.Empty;
			Image = null;
			Name = "";
			Parent = null;
			Matrix = new QColorMatrix();
			Matrix.Reset();
			Alpha = 255;
			FlipFlags = EFlipFlag.FlipNone;
			Inverted = false;
			FlippedImage = null;

			SubLayers = new List<RPGLayer>();
			MatrixValues = new float[4] { 0, 1, 1, 0 };
		}

		public void SetImage(RPGPictureBox Img)
		{
			if (Img == null)
			{
				Image = null;
				Program.layersWindowForm.PictureBox1.Image = null;
				FlippedImage = null;
				return;
			}

			Image = Img.MainImage;
			Program.layersWindowForm.PictureBox1.Image = Img.MyImage;
			UpdateFlippedImage();
		}

		public void UpdateFlippedImage()
		{
			if (Image == null)
				return;

			FlippedImage = Image.RawBitmap.Clone(new Rectangle(0, 0, Image.RawBitmap.Width, Image.RawBitmap.Height), System.Drawing.Imaging.PixelFormat.Format32bppArgb);
			if ((FlipFlags & (EFlipFlag.FlipHorizontal | EFlipFlag.FlipVertical)) == (EFlipFlag.FlipHorizontal | EFlipFlag.FlipVertical))
				FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipXY);
			else if ((FlipFlags & EFlipFlag.FlipHorizontal) == EFlipFlag.FlipHorizontal)
				FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipX);
			else if ((FlipFlags & EFlipFlag.FlipVertical) == EFlipFlag.FlipVertical)
				FlippedImage.RotateFlip(RotateFlipType.RotateNoneFlipY);
		}

		public void Save(System.IO.BinaryWriter File, bool ImagesToo)
		{
			File.Write(Offset.X);
			File.Write(Offset.Y);
			File.Write(Name);
			File.Write(Image != null);

			if (Image != null)
			{
				File.Write(Image.Path);

				if (ImagesToo)
				{
					File.Write(Image.OriginalBitmap.Width);
					File.Write(Image.OriginalBitmap.Height);

					Canvas.WriteBitmapToFile(File, Image.OriginalBitmap);
				}
			}

			for (int x = 0; x <= 4; x++)
			{
				for (int y = 0; y <= 4; y++)
					File.Write(Matrix.MyMatrix[x, y]);
			}

			for (int i = 0; i <= 3; i++)
				File.Write(MatrixValues[i]);

			File.Write(Alpha);
			File.Write((int)FlipFlags);
			File.Write(Inverted);

			// Parent is not saved/loaded, it is re-created on run-time
			File.Write(SubLayers.Count);

			foreach (var SubLayer in SubLayers)
				SubLayer.Save(File, ImagesToo);
		}

		public void Load(System.IO.BinaryReader File, bool ImagesToo)
		{
			Offset = new Point(File.ReadInt32(), File.ReadInt32());
			Name = File.ReadString();
			if (File.ReadBoolean() == true)
			{
				string Img = File.ReadString();
				Image = Images.FindImage(Img);

				if (Image != null)
					FlippedImage = Image.RawBitmap;

				if (ImagesToo)
				{
					int Width = File.ReadInt32();
					int Height = File.ReadInt32();

					using (Bitmap BMP = new Bitmap(Width, Height, System.Drawing.Imaging.PixelFormat.Format32bppArgb))
					{
						Canvas.ReadBitmapFromFile(File, Width, Height, BMP);
						// Do we have this image
						if (System.IO.File.Exists(Img))
						{
							if (Images.BitmapCompare(BMP, Images.FindImage(Img).OriginalBitmap) == false)
							{
								// Message the user
								if (MessageBox.Show("The image " + Img + " already exists in your image folder, but yours differs from the one in the file. Would you like to overwrite with this one?", "", MessageBoxButtons.YesNo) == DialogResult.Yes)
								{
									BMP.Save(Img);
									Program.canvasForm.ImagesAcquired.Add(Img);
								}
							}
						}
						else if (System.IO.File.Exists(Img) == false)
						{
							System.IO.Directory.CreateDirectory(Img.Substring(0, Img.LastIndexOf("\\")));

							// Save the bitmap
							BMP.Save(Img);
							Program.canvasForm.ImagesAcquired.Add(Img);

							// Set our image to a temporary image for later re-loading
							Image = new RPGImage();
							Image.Path = Img;
						}
					}

				}
				else
				{
					if (Image == null)
						Program.canvasForm.PackImagesNotFound.Add(Img);
				}
			}

			for (int x = 0; x <= 4; x++)
			{
				for (int y = 0; y <= 4; y++)
					Matrix.MyMatrix[x, y] = File.ReadSingle();
			}

			for (int i = 0; i <= 3; i++)
				MatrixValues[i] = File.ReadSingle();

			Alpha = File.ReadInt32();
			FlipFlags = (EFlipFlag)File.ReadInt32();
			Inverted = File.ReadBoolean();

			int SubLayerNum = File.ReadInt32();
			for (int i = 0; i <= SubLayerNum - 1; i++)
			{
				RPGLayer SubLayer = new RPGLayer();
				SubLayer.Load(File, ImagesToo);
				SubLayer.Parent = this;
				SubLayers.Add(SubLayer);
			}
		}

		public System.Drawing.Imaging.ImageAttributes GenerateColorMatrix()
		{
			System.Drawing.Imaging.ImageAttributes Attr = new System.Drawing.Imaging.ImageAttributes();
			Attr.SetColorMatrix(Matrix.MyMatrix, System.Drawing.Imaging.ColorMatrixFlag.Default, System.Drawing.Imaging.ColorAdjustType.Bitmap);
			return Attr;
		}

		public void ReloadImagesSubLayers(bool Restore)
		{
			if (Restore == false)
			{
				foreach (var SubLayer in SubLayers)
				{
					if (SubLayer.Image != null)
					{
						RPGImage Img = SubLayer.Image;

						SubLayer.Image = new RPGImage();
						SubLayer.Image.Path = Img.Path;
					}
					SubLayer.ReloadImagesSubLayers(Restore);
				}
			}
			else
			{
				foreach (var SubLayer in SubLayers)
				{
					if (SubLayer.Image != null)
					{
						SubLayer.Image = Images.FindImage(SubLayer.Image.Path);
						SubLayer.UpdateFlippedImage();
					}
					SubLayer.ReloadImagesSubLayers(Restore);
				}
			}
		}

		public void SetColorMatrix()
		{
			Matrix.Reset();
			Matrix.ScaleColors(float.Parse(Program.layersWindowForm.TextBox4.Text));
			Matrix.SetSaturation(float.Parse(Program.layersWindowForm.TextBox3.Text));
			Matrix.RotateHue(int.Parse(Program.layersWindowForm.TextBox2.Text));
			Matrix.ScaleOpacity((float)(Alpha / 255.0f));

			if (Inverted)
				Matrix.Invert();

			Matrix.TranslateColors(float.Parse(Program.layersWindowForm.TextBox6.Text));
			Program.canvasForm.UpdateDrawing();
		}

		// CANVAS RENDERING
		public void RenderCanvasLayer(Graphics gfx)
		{
			float Zoom = (float)(Program.canvasForm.ZoomNumber) / 100.0f;
			int ContentWidth = Program.canvasForm.ToolStripContainer1.ContentPanel.Width / 2;
			int ContentHeight = Program.canvasForm.ToolStripContainer1.ContentPanel.Height / 2;
			Bitmap bmp = FlippedImage;
			Rectangle destRect;
			Rectangle srcRect;
			System.Drawing.Imaging.ImageAttributes Attr = GenerateColorMatrix();

			if (bmp == null)
				return;

			// Frame mode
			if (Program.canvasForm.FrameAnimationToolStripMenuItem.Checked)
			{
				int BmpWidth = (int)(((bmp.Width / Image.GameFile.SheetRows) * Zoom) / 2);
				int BmpHeight = (int)(((bmp.Height / Image.GameFile.SheetColumns) * Zoom) / 2);
				int DrawX = (int)((ContentWidth - BmpWidth) + (Offset.X * Zoom));
				int DrawY = (int)((ContentHeight - BmpHeight) + (Offset.Y * Zoom));
				int Frame;

				if (Program.canvasForm.AnimationIndex == 0)
					Frame = 0;
				else
					Frame = (int)(Images.CurrentGameFile.Animations[Program.canvasForm.AnimationIndex - 1].GetFrame(Program.canvasForm.AnimationFrame));

				destRect = new Rectangle(DrawX, DrawY, (int)((bmp.Width / Image.GameFile.SheetRows) * Zoom), (int)((bmp.Height / Image.GameFile.SheetColumns) * Zoom));
				srcRect = Character.GetRectangleForBitmapFrame(bmp, Frame, Image.GameFile.SheetRows, Image.GameFile.SheetColumns);
				gfx.DrawImage(bmp, destRect, srcRect.X, srcRect.Y, srcRect.Width, srcRect.Height, GraphicsUnit.Pixel, Attr);
			}
			else if (Program.canvasForm.SingleSetToolStripMenuItem.Checked) // Sheet mode
			{
				int BmpWidth = (int)((bmp.Width * Zoom) / 2);
				int BmpHeight = (int)((bmp.Height * Zoom) / 2);
				int DrawX = (int)((ContentWidth - BmpWidth) + (Offset.X * Zoom) + Program.canvasForm.CameraPosition.X);
				int DrawY = (int)((ContentHeight - BmpHeight) + (Offset.Y * Zoom) + Program.canvasForm.CameraPosition.Y);

				destRect = new Rectangle(new Point(DrawX, DrawY), new Size((int)(bmp.Width * Zoom), (int)(bmp.Height * Zoom)));
				srcRect = new Rectangle(0, 0, bmp.Width, bmp.Height);

				gfx.DrawImage(bmp, destRect, srcRect.X, srcRect.Y, srcRect.Width, srcRect.Height, GraphicsUnit.Pixel, Attr);
			}
			else
			{
				int BmpWidth = (int)(bmp.Width / 2);
				int BmpHeight = (int)(bmp.Height / 2);
				ContentWidth = (int)(gfx.VisibleClipBounds.Width / 2);
				ContentHeight = (int)(gfx.VisibleClipBounds.Height / 2);
				int DrawX = (int)((ContentWidth - BmpWidth) + (Offset.X));
				int DrawY = (int)((ContentHeight - BmpHeight) + (Offset.Y));

				destRect = new Rectangle(new Point(DrawX, DrawY), new Size(bmp.Width, bmp.Height));
				srcRect = new Rectangle(0, 0, bmp.Width, bmp.Height);

				gfx.DrawImage(bmp, destRect, srcRect.X, srcRect.Y, srcRect.Width, srcRect.Height, GraphicsUnit.Pixel, Attr);
			}
		}

		public void DrawCanvasLayers(Graphics e)
		{
			if (Image != null)
				RenderCanvasLayer(e);

			foreach (var SubLayer in SubLayers)
				SubLayer.DrawCanvasLayers(e);
		}

		public void RenderLayerToGraphics(Graphics gfx)
		{
			var WImg = OwnedCharacter.SizeOfOutput;
			Size ContentSize = new Size(WImg.Width / 2, WImg.Height / 2);
			Size BmpSize = new Size(FlippedImage.Width / 2, FlippedImage.Height / 2);
			Point DrawLocation = new Point((ContentSize.Width - BmpSize.Width) + Offset.X, (ContentSize.Height - BmpSize.Height) + Offset.Y);

			gfx.DrawImage(FlippedImage, new Rectangle(DrawLocation, FlippedImage.Size), 0, 0, FlippedImage.Width, FlippedImage.Height, GraphicsUnit.Pixel, GenerateColorMatrix());
		}

		public void RenderLayerToGraphicsRecursion(Graphics e)
		{
			if (Image != null)
				RenderLayerToGraphics(e);

			foreach (var SubLayer in SubLayers)
				SubLayer.RenderLayerToGraphicsRecursion(e);
		}
	}

	public class RPGCharacter
	{
		public string Name { get; set; } // Unique ID for character file (for multi-sheet)
		public List<RPGLayer> Layers { get; private set; }

		public RPGCharacter()
		{
			Layers = new List<RPGLayer>();
		}

		public void Save(System.IO.BinaryWriter Writer, bool ImagesToo)
		{
			Writer.Write(Layers.Count);

			foreach (var Layer in Layers)
				Layer.Save(Writer, ImagesToo);
		}

		public void Load(System.IO.BinaryReader Reader, bool ImagesToo)
		{
			int Count = Reader.ReadInt32();

			for (int i = 0; i <= Count - 1; i++)
			{
				RPGLayer NewLayer = new RPGLayer();
				NewLayer.Load(Reader, ImagesToo);
				Layers.Add(NewLayer);
			}

			foreach (var Layer in Layers)
			{
				RPGNode Node = new RPGNode();
				Node.Layer = Layer;
				Node.Text = Layer.Name;
				Layer.Node = Node;

				Program.layersWindowForm.TreeView1.Nodes.Add(Node);

				Node.RecurseAndLoadSubNodes();
			}
		}

		private Size SizeRecursion(List<RPGLayer> LayerCollection, Size CurrentSize)
		{
			foreach (var Layer in LayerCollection)
			{
				CurrentSize = SizeRecursion(Layer.SubLayers, CurrentSize);

				if (Layer.FlippedImage == null)
					continue;

				if (Layer.FlippedImage.Size.Width > CurrentSize.Width && Layer.FlippedImage.Size.Height > CurrentSize.Height)
					CurrentSize = Layer.FlippedImage.Size;
			}

			return CurrentSize;
		}

		public Size SizeOfOutput
		{
			get
			{
				if (Layers.Count == 0)
					return Size.Empty;

				Size LargestSize = new Size(0, 0);
				return SizeRecursion(Layers, LargestSize);
			}
		}

		public RPGLayer CreateLayer(TreeNode Node)
		{
			RPGLayer Layer = new RPGLayer();

			if ((Node != null) && (Node.Parent != null))
			{
				// Use our parent layer
				Layer.Parent = ((RPGNode)Node.Parent).Layer;
				Layer.OwnedCharacter = Layer.Parent.OwnedCharacter;
				Layer.Parent.SubLayers.Add(Layer);
			}
			else
			{
				Layer.Parent = null;
				Layer.OwnedCharacter = this;
				Layers.Add(Layer);
			}

			Layer.Name = Node.Text;
			return Layer;
		}

		public void RemoveLayer(RPGLayer Layer)
		{
			// Do we have a parent
			if (Layer.Parent != null)
				Layer.Parent.SubLayers.Remove(Layer); // Remove us from sublayers
			else
				Layers.Remove(Layer);
		}

		public void MoveLayer(RPGLayer Layer, bool Down)
		{
			if (Layer.Parent != null)
			{
				int OldIndex = (int)(Layer.Parent.SubLayers.IndexOf(Layer));
				Layer.Parent.SubLayers.Remove(Layer);

				if (Down)
					Layer.Parent.SubLayers.Insert(OldIndex + 1, Layer);
				else
					Layer.Parent.SubLayers.Insert(OldIndex - 1, Layer);
			}
			else
			{
				int OldIndex = (int)(Program.characterSelectForm.GetSelectedCharacter().Character.Layers.IndexOf(Layer));
				Layers.Remove(Layer);

				if (Down)
					Layers.Insert(OldIndex + 1, Layer);
				else
					Layers.Insert(OldIndex - 1, Layer);
			}
			Program.canvasForm.UpdateDrawing();
		}
	}

	public class Character
	{
		public static Rectangle GetRectangleForBitmapFrame(Image bmp, int FrameNumber, int NumRows, int NumColumns)
		{
			if (bmp == null)
				return new Rectangle(0, 0, 0, 0);

			int SpriteWidth = bmp.Width / NumRows;
			int SpriteHeight = bmp.Height / NumColumns;
			int FramesPerRow = bmp.Width / SpriteWidth;

			return new Rectangle(((FrameNumber % FramesPerRow) * SpriteWidth), ((FrameNumber / FramesPerRow) * SpriteHeight), SpriteWidth, SpriteHeight);
		}
	}
}
